nidx/src/searcher/grpc.rs

```diff
@@ -18,79 +18,166 @@
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 //
 
+use std::collections::HashMap;
 use std::{pin::Pin, sync::Arc};
 
 use futures::Stream;
+use nidx_protos::nidx::nidx_searcher_client::NidxSearcherClient;
 use nidx_protos::nidx::nidx_searcher_server::{NidxSearcher, NidxSearcherServer};
 use nidx_protos::*;
+use tokio::sync::RwLock;
+use tonic::transport::Channel;
 use tonic::{service::Routes, Request, Response, Result, Status};
 
-use super::streams::{document_iterator, paragraph_iterator};
+use crate::errors::{NidxError, NidxResult};
+use crate::searcher::shard_selector::SearcherNode;
+
+use super::shard_selector::ShardSelector;
+use super::streams;
 use super::{index_cache::IndexCache, shard_search, shard_suggest};
 use tracing::*;
 
 pub struct SearchServer {
     index_cache: Arc<IndexCache>,
+    shard_selector: ShardSelector,
+    clients: Arc<RwLock<HashMap<String, NidxSearcherClient<Channel>>>>,
 }
 
 impl SearchServer {
-    pub fn new(index_cache: Arc<IndexCache>) -> Self {
+    pub fn new(index_cache: Arc<IndexCache>, shard_selector: ShardSelector) -> Self {
         SearchServer {
             index_cache,
+            shard_selector,
+            clients: Arc::new(RwLock::new(HashMap::new())),
         }
     }
 
     pub fn into_service(self) -> Routes {
         Routes::new(NidxSearcherServer::new(self))
     }
+
+    async fn get_client(&self, hostname: &String) -> NidxResult<NidxSearcherClient<Channel>> {
+        if let Some(client) = self.clients.read().await.get(hostname) {
+            return Ok(client.clone());
+        }
+
+        let client = NidxSearcherClient::connect(format!("http://{hostname}")).await?;
+        self.clients.write().await.insert(hostname.clone(), client.clone());
+        Ok(client)
+    }
+}
+
+/// Macro to run a request on a shard, local or remote
+macro_rules! shard_request {
```

Comment:
Since you are calling different functions depending on the endpoint, you cannot do this with functions (because they are async functions and the types are opaque). Or at least, I was not able to.

The other option that I was able to make work is to copy-paste this block into each function implementation.